### PR TITLE
Use control requests for updating the bootloader

### PR DESCRIPTION
**Pull request by sergeuz**
## Description

This PR updates the `update` command so that it can flash certain modules of the target release, such as the bootloader module, using control requests. Consequently, the `update` command no longer requires the target device to be in DFU mode and instead automatically switches it to an appropriate state before flashing.

Other changes in this PR:

- Ensure there's a single place where the list of supported platforms is defined (see platform.js). Previously, different commands/subsystems of the CLI supported different sets of platforms due to the definitions from device-constants being filtered differently in those subsystems. Consequently, the remaining support for the following platforms was removed in this PR: asom, xsom, raspi, oak, duo, bluz.
- Change the order in which the firmware modules are updated so that the bootloader is flashed first.
- Use the device ID to identify the device when flashing it via `dfu-util`.

## How to Test

- Validate that the `update` and `flash` commands work as expected.
- Run unit tests.
- Run end-to-end tests.

## Related Issues / Discussions

- [sc-104580](https://app.shortcut.com/particle/story/104580/particle-update-command-cannot-be-used-to-update-bootloader-on-p2)

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

**Comment by monkbroc** [comment]
`particle update` is the main method to bring devices that are running old versions of Device OS up to date. I think the upgrade path proposed won't work on Device OS versions that don't support control requests.

I tested with an Electron running Device OS 1.0.1 that was running some unknown application in DFU mode, and the device went back to normal mode and eventually the CLI gave a timeout.
```
$ particle update

> Your device is ready for a system update.
> This process may take a few minutes. Here it goes!

! An error occurred while attempting to update the system firmware of your device:

TimeoutError: Request timeout

> Please visit our community forums for help with this error:
https://community.particle.io/
```

```
particle serial inspect

Platform: 10 - Electron
Modules
  Bootloader module #0 - version 201, main location, 16384 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #1 - version 1002, main location, 131072 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #3 - version 1002
  System module #2 - version 1002, main location, 131072 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #1 - version 1002
      Bootloader module #0 - version 201
  System module #3 - version 1002, main location, 131072 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #2 - version 110
  User module #1 - version 5, main location, 131072 bytes max size
    UUID: 2FBE7D89911634323A38F8D56E26F16F75DFAA7D51F0D2C889B6464F69E84998
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #2 - version 1002
  empty - factory location, 131072 bytes max size
```

Sergey, is there a path forward to ensure that older devices can be upgraded? Is flashing the bootloader first mandatory?

**Comment by monkbroc** [comment]
I tried with the same Electron in listening mode. The first time the LED changed to magenta and the CLI hung forever until I unplugged the USB cable then I got `UsbError: IN control transfer failed`. I plugged the device back and put it in listening mode and the second time the CLI exited right away with `UsbError: OUT control transfer failed`.

With the device in normal mode I get `TimeoutError: Request timeout`

**Comment by monkbroc** [comment]
For a fresh out of the box Argon, the CLI exits right away with `No devices found` due to an issue with the serial port not being configured properly in an older 0.8.0-rc (I can't remember which one).

**Comment by monkbroc** [comment]
Putting the system firmware first in `settings.js` made the `particle update` work for my Electron 1.0.1 and both of my old Argons on 0.8.0-rc.X.

**Review by monkbroc — CHANGES_REQUESTED** [review]
The approach is not reliable enough to replace `particle update` but worked well when the bootloader was flashed last.

**Comment by sergeuz** [comment]
> Is flashing the bootloader first mandatory?

@monkbroc The rationale was that flashing it first is _safer_ as we've had issues in the past caused by various incompatibilities between old bootloader and newer system part modules. That's also how devices are updated OTA, so when we introduce changes in Device OS that affect both the bootloader and system firmware, it's assumed that the bootloader will be updated first.

I thought that early Gen 3 devices that come with 0.8.0-rc.X flashed on them still need to be updated using the mobile apps first, but if the CLI has to support updating those devices too, reversing the order is the way to go :+1:

**Comment by sergeuz** [comment]
@monkbroc Could you give the last [patch](https://github.com/particle-iot/particle-cli/pull/623/commits/fa077429be4380dfc22063f4d3fdc1d26d7b22c5) a try? I tested it with an Argon running 0.8.0-rc.13 (IIRC, that was the first manufacturing release for Gen 3) and an Electron running 0.5.4 and it worked fine.

The logic is the following:

- If the device is in DFU mode when `particle update` is executed, the modules that need to be flashed to the device are reordered so that the modules that can be flashed via DFU are flashed first. That ensures that the device will have an updated system firmware before we attempt to flash the remaining modules via control requests. 
- If the device is not in DFU mode and is running an older version of Device OS than 2.0.0, we fail and suggest the user to switch the device into DFU mode.
- Devices running Device OS 2.0.0 and higher are flashed following the original logic from this PR.

File: src/cmd/update.js
Comment (monkbroc) [inline src/cmd/update.js:66]: It's not necessary to give this information to the user. The next line (please switch to DFU) is sufficient.

**Review by monkbroc — APPROVED** [review]
The new approach worked for me. I'm approving the PR. I'd like Mirande to take a look as well next week when he's back before this is merged and released.

**Review by busticated — COMMENTED** [review]
awesome thanks @sergeuz 🙏 

i'm seeing some weirdness in the `e2e` tests but i'm still poking around + testing here

File: package.json
Comment (sergeuz) [reply package.json:55]: Yeah, I planned to publish the changes from [device-constants/#22](https://github.com/particle-iot-inc/device-constants/pull/22) first and then update it here. Could you give that PR a look as well?

**Comment by busticated** [comment]
i'm seeing a bunch of failures running the `e2e` tests locally with a device connected. seems like these steps are no longer working as expected (timing out for some reason):

https://github.com/particle-iot/particle-cli/blob/control-req-update/sc-104580/test/e2e/flash.e2e.js#L45-L46

i'm running `npm run test:e2e -- --grep="Flash Commands"` - are others able to run those successfully? 

fwiw, my device info below:

<details>
<summary><b>`particle serial inspect` output</b></summary>
<p>

```
Platform: 12 - Argon
Modules
  Bootloader module #0 - version 1006, main location, 49152 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #1 - version 2301, main location, 671744 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      Bootloader module #0 - version 1005
      Radio stack module #0 - version 202
  User module #1 - version 6, main location, 131072 bytes max size
    UUID: 6EDEE61AAF0801707C245DBBB97973588BAE77A8127A210FE5A7A278AB812C0A
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #1 - version 2301
  NCP module #0 - version 5, main location, 1536000 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  Radio stack module #0 - version 202, main location, 192512 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
```

</p>
</details>

**Comment by sergeuz** [comment]
> i'm running npm run test:e2e -- --grep="Flash Commands" - are others able to run those successfully?

Oops, I forgot to change the hardcoded device name in one of those tests. The tests pass for me reliably though on an Argon running the same firmware as yours :shrug: The code paths of the `flash` and `usb` commands are largely unchanged in this PR, so I'm puzzled as to what might be causing the problem in your case.

**Comment by busticated** [comment]
@sergeuz more info on my `e2e` issues. seems i'm hitting errors trying to query for cloud status after resetting the device.

the command (`particle usb cloud-status <device-id> --until connected`) basically loops recursively ignoring errors until it either gets the expected status or times-out. when i log out debug info i see a series of errors like:

```
The device should not be in DFU mode
Failed to set alt setting
Missing permissions to access the USB device
Failed to claim interface
Unable to open USB device
Unable to connect to the device <id>. Make sure the device is connected to the host computer via USB
Unable to open USB device
Unable to enumerate USB devices
```

>The tests pass for me reliably though on an Argon running the same firmware as yours

ah, ok great. that seems a stronger hint that something is weird in my local env. i'll keep an eye on it but i think this is good to ship.

edit: ah, CI is now flaking out w/ `failed to create host: Image xcode:13 is not supported` - i'll get a PR up for that ASAP as it'll block releasing :+1:

**Review by busticated — APPROVED** [review]

**Comment by sergeuz** [comment]
> seems i'm hitting errors trying to query for cloud status after resetting the device.

@busticated Do those tests pass in `master` for you?

> CI is now flaking out w/ failed to create host: Image xcode:13 is not supported - i'll get a PR up for that ASAP

Thank you!

**Comment by sergeuz** [comment]
@busticated Do you see the same issue in `master` or is it only reproducible with this branch for you?

**Comment by busticated** [comment]
@sergeuz i went back and tested previous versions - the issue reproduces just by running `npm run test:e2e -- --grep="USB DFU Subcommand" --bail` - some more info:

* `v3.2.0`: fail (`particle-usb@2.0.0` | [source](https://github.com/particle-iot/particle-cli/tree/cbb0b08645a96bbe22d4715c1b502818c55acab6))
* `v3.1.0`: fail (`particle-usb@2.0.0` | [source](https://github.com/particle-iot/particle-cli/tree/aa860e6e2843fe9a0984588c8ecda863983bfc8e))
* `v3.0.0`: fail (`particle-usb@2.0.0` | [source](https://github.com/particle-iot/particle-cli/tree/329a14a4c910cfb2d2d2b323152e732c73321c59)) 
* `v2.16.0`: ok (`particle-usb@1.2.1` | [source](https://github.com/particle-iot/particle-cli/tree/7e642afda6bb0fde65057a00fb3cd0b734aca159)) 

in the case of `v3.1.0`, i saw it pass _once_ but then re-runs failed multiple times. `v2.16.0` passed reliably. so, i guess i have some more digging to do 😅 😬 😞 

i don't think that needs to hold up merging this PR since this issue obviously predates your changes.

thanks for the help 🙏 👍

**Comment by sergeuz** [comment]
@busticated The only relevant change I noticed between 2.16.0 and 3.0.0 is that we updated particle-usb to 2.0.0 in the latter release :thinking: 

I'm going to merge this PR given that the issue is not new. Let's create a story to investigate it further.

**Pull request merged**
